cpp/src/io/csv/csv_reader_impl.cu

```diff
@@ -698,7 +716,32 @@ table reader::Impl::read()
 
   // Transfer ownership to raw pointer output arguments
   std::vector<gdf_column *> out_cols(num_active_cols);
+
+  auto maybe_cast_datetimes = args_.out_time_unit != TIME_UNIT_NONE;
+
   for (size_t i = 0; i < out_cols.size(); ++i) {
+    if (maybe_cast_datetimes) {
+
+      auto is_datetime = columns[i]->dtype == GDF_DATE32 ||
+                         columns[i]->dtype == GDF_DATE64 ||
+                         columns[i]->dtype == GDF_TIMESTAMP;
+
+      if (is_datetime && columns[i]->dtype_info.time_unit != args_.out_time_unit) {
+        // Cast the datetime-like column to the desired out_time_unit
+        auto col = columns[i].get();
+        gdf_dtype_extra_info dtype_info{args_.out_time_unit};
+        auto res = cudf::cast(*col, GDF_TIMESTAMP, dtype_info);
```

Comment:
We would want to do in-kernel conversion; basically modify the date/timestamp parsing functions to convert to the desired output time unit, which is assumed to be milliseconds right now.
We will be doing refactoring of the CSV reader though so we can do it as part of that enhancement.

Reply:
Yes that was the path I'd gone down originally, glad to hear a more wholistic refactor is planned.

Reply:
Agree in-kernel conversion would be best and necessary if the parser is capable of loading a higher resolution than milliseconds.  It doesn't look like that's the case today, so this should be fine for now.